sympy/core/expr.py

```diff
@@ -329,98 +329,66 @@ def __complex__(self):
         re, im = result.as_real_imag()
         return complex(float(re), float(im))
 
-    def __ge__(self, other):
-        from sympy import GreaterThan
+    def _cmp(self, other, op, cls):
+        assert op in ("<", ">", "<=", ">=")
         try:
             other = _sympify(other)
         except SympifyError:
-            raise TypeError("Invalid comparison %s >= %s" % (self, other))
+            raise TypeError("Invalid comparison %s %s %s" % (self, op, other))
         for me in (self, other):
             if me.is_extended_real is False:
                 raise TypeError("Invalid comparison of non-real %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
+
         n2 = _n2(self, other)
         if n2 is not None:
-            return _sympify(n2 >= 0)
+            if op == "<":
+                return _sympify(n2 < 0)
+            elif op == ">":
+                return _sympify(n2 > 0)
+            elif op == "<=":
+                return _sympify(n2 <= 0)
+            else: # >=
+                return _sympify(n2 >= 0)
+
         if self.is_extended_real and other.is_extended_real:
-            if (self.is_infinite and self.is_extended_positive) \
-                    or (other.is_infinite and other.is_extended_negative):
-                return S.true
-            nneg = (self - other).is_extended_nonnegative
-            if nneg is not None:
-                return sympify(nneg)
-        return GreaterThan(self, other, evaluate=False)
+            # this should work for infinite values as well,
+            # so we don't need a separate check
+            diff = self - other
+            if diff is not S.NaN:
+                if op == "<":
+                    test = sympify(diff.is_extended_negative)
+                elif op == ">":
+                    test = sympify(diff.is_extended_positive)
+                elif op == "<=":
+                    test = sympify(diff.is_extended_nonpositive)
+                else: # >=
+                    test = sympify(diff.is_extended_nonnegative)
+
+                if test is not None:
+                    return sympify(test)
+            else: # e.g. oo > oo
+                raise ValueError("Undefined comparison of infinite values")
```

Comment:
What happens for `oo <= oo`?

Reply:
`oo <= oo` is equivalent to `oo - oo <= 0` which should result in the ValueError being raised

Reply:
I will see about adding a few tests for code coverage

Reply:
I suppose treating `oo <= oo` as an undefined expression is different than the current implementation, but I feel like it is the correct thing to do. Mathematically, `oo` represents the limit of some real-valued thing, so the value of comparing two infinities of the same sign is undefined unless we know something about how fast each thing grows in the limit

Reply:
We are using the extended real numbers so we should have `oo <= oo`:
https://en.wikipedia.org/wiki/Extended_real_number_line#Order_and_topological_properties

You need to be clear what this PR is doing: are you just reorganising code or are you changing the definition of `<=`?

Reply:
On master we have:
```julia
In [1]: x = Symbol('x', extended_positive=True, infinite=True)                                                                                                

In [2]: y = Symbol('y', extended_positive=True, infinite=True)                                                                                                

In [3]: x <= y                                                                                                                                                
Out[3]: True

In [4]: Le(x, y)                                                                                                                                              
Out[4]: True
```
This PR leaves those two unevaluated.

Reply:
Yeah, I decided to change it back. You're right, it is sometimes helpful to define it that way

Reply:
Alright, it should work like that now